### PR TITLE
support multiple connections on multiplex conn pool

### PR DESCRIPTION
**Pull request by cch123**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by cch123** [comment]
use dubbo to test:

File: pkg/stream/xprotocol/connpool_multiplex.go [hunk 1]
```diff
@@ -59,24 +69,30 @@ func (p *poolMultiplex) init(client *activeClientMultiplex, sub types.ProtocolNa
...
-			p.activeClients.Delete(sub)
+			p.activeClients[index].Delete(sub)
 		}
 	}, nil)
 }
 
 // CheckAndInit init the connection pool
 func (p *poolMultiplex) CheckAndInit(ctx context.Context) bool {
 	var client *activeClientMultiplex
+	clientIdx := atomic.AddInt64(&p.currentCheckAndInitIdx, 1) % int64(len(p.activeClients))
```
Comment (taoyuanyuan) [inline pkg/stream/xprotocol/connpool_multiplex.go:83]: a request maybe call the func more than once

Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go:83]: @taoyuanyuan , this is ok, the caller only return the pool when CheckAndInit returns true, and under this situation, the client slot must be inited and the index is saved to the ctx

Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go:82]: currently, we'll set the client index to context when CheckAndInit returns true, and use this index to find the client in the pool should be valid

File: pkg/stream/xprotocol/connpool_multiplex.go [hunk 2]
```diff
@@ -136,14 +159,17 @@ func (p *poolMultiplex) NewStream(ctx context.Context, receiver types.StreamRece
 
 // Shutdown stop the keepalive, so the connection will be idle after requests finished
 func (p *poolMultiplex) Shutdown() {
-	f := func(k, v interface{}) bool {
-		ac, _ := v.(*activeClientMultiplex)
-		if ac.keepAlive != nil {
-			ac.keepAlive.keepAlive.Stop()
+	atomic.StoreInt64(&p.shutdown, 1)
+	for i := 0; i < len(p.activeClients); i++ {
```
Comment (taoyuanyuan) [inline pkg/stream/xprotocol/connpool_multiplex.go:172]: for _, client := range p.activeclient

Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go:172]: image

This grammar will cause the lock in sync.Map to be copied

File: pkg/stream/xprotocol/connpool_multiplex.go [hunk 3]
```diff
@@ -36,47 +36,72 @@ import (
...
 	}, nil)
 }
 
 // CheckAndInit init the connection pool
 func (p *poolMultiplex) CheckAndInit(ctx context.Context) bool {
-	var client *activeClientMultiplex
+	var clientIdx int64 = 0 // most use cases, there will only be 1 connection
+	if len(p.activeClients) > 1 {
+		if clientIdx = getClientIDFromDownStreamCtx(ctx); clientIdx == invalidClientID {
+			clientIdx = atomic.AddInt64(&p.currentCheckAndInitIdx, 1) % int64(len(p.activeClients))
```
Comment (wangfakang) [inline pkg/stream/xprotocol/connpool_multiplex.go:91]: maybe better use `AddUint64`  ? because the value type of `Connections().Max()`  is `uint64`.

Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go:91]: invalidClientID == -1 is used to check whether the client id is set in context

File: pkg/stream/xprotocol/connpool_multiplex.go [hunk 4]
```diff
@@ -136,14 +174,26 @@ func (p *poolMultiplex) NewStream(ctx context.Context, receiver types.StreamRece
...
 // Shutdown stop the keepalive, so the connection will be idle after requests finished
 func (p *poolMultiplex) Shutdown() {
-	f := func(k, v interface{}) bool {
-		ac, _ := v.(*activeClientMultiplex)
-		if ac.keepAlive != nil {
-			ac.keepAlive.keepAlive.Stop()
+	utils.GoWithRecover(func() {
+		p.clientMux.Lock()
+		if p.shutdown {
+			return
```
Comment (nejisama) [inline pkg/stream/xprotocol/connpool_multiplex.go:180]: return without unlock?

Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go]: good catch

File: pkg/stream/xprotocol/connpool_multiplex.go [hunk 5]
```diff
@@ -87,19 +112,32 @@ func (p *poolMultiplex) CheckAndInit(ctx context.Context) bool {
...
+			// this client is not inited
+			return p.Host(), nil, types.ConnectionFailure
+		}
+	}
+
 	subProtocol := getSubProtocol(ctx)
 
-	client, _ := p.activeClients.Load(subProtocol)
-	host := p.Host()
+	client, _ := p.activeClients[clientIdx].Load(subProtocol)
```
[the remark this answers is not in the document]
Comment (cch123) [reply pkg/stream/xprotocol/connpool_multiplex.go:138]: only when CheckAndInit returns true, the getActiveConnectionPool returns this pool, so the clientIdx must be a valid value

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Pull request merged**